xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-link/plugin.js

```diff
@@ -111,6 +111,139 @@
+
+            require([
+              'xwiki-attachments-store',
+              'xwiki-file-picker',
+              'resource'
+            ], function (attachmentsStore, filePicker, resource) {
+              const convertFilesToAttachments = function (files, documentReference) {
+                const attachments = [];
+                for (var i = 0; i < files.length; i++) {
+                  const file = files.item(i);
+                  const attachmentReference = new XWiki.EntityReference(file.name, XWiki.EntityType.ATTACHMENT,
+                    documentReference);
+                  attachments.push(attachmentsStore.create(attachmentReference, file));
+                }
+                return attachments;
+              };
+
+              // Open the file picker
+              filePicker.pickLocalFiles({
+                accept: "*",
+                multiple: false
+              }).then(function (files) {
+                const attachments = convertFilesToAttachments(
+                  files,
+                  editor.config.sourceDocument.documentReference
+                );
+
+                // Cancel the insertion when no file is picked.
+                if (attachments.length === 0) {
+                  return;
+                }
+
+                const attachment = attachments[0];
+                const resourceReference = resource.convertEntityReferenceToResourceReference(
+                  XWiki.Model.resolve(attachment.id, XWiki.EntityType.ATTACHMENT),
+                  editor.config.sourceDocument.documentReference);
+
+                // Insert the link immediately after closing the file picker.
+                editor.insertHtml($('<a></a>').attr({
+                    href: CKEDITOR.plugins.xwikiResource.getResourceURL(resourceReference, editor),
+                    'data-reference': CKEDITOR.plugins.xwikiResource.serializeResourceReference(resourceReference)
+                  }).text(attachment.name).prop('outerHTML'));
+
+                const notification = new XWiki.widgets.Notification(
+                  editor.localization.get('xwiki-link.uploadProgress', attachment.name),
+                  'inprogress');
+
+                // Upload the selected attachment
+                const attachmentReference = XWiki.Model.resolve(attachment.id, XWiki.EntityType.ATTACHMENT);
+                attachmentsStore.upload(attachmentReference, attachment.file).then(() => {
+                  notification.replace(
+                    new XWiki.widgets.Notification(
+                      editor.localization.get('xwiki-link.uploadSuccess', attachment.name),
+                      'done')
+                  );
+                }).catch(() => {
+                  notification.replace(
+                    new XWiki.widgets.Notification(
+                      editor.localization.get('xwiki-link.uploadError', attachment.name),
+                      'error')
+                  );
+                  return Promise.reject();
```

Comment:
Is this needed? You don't chain anything after the ``catch`` call, and you don't return the promise either.